### PR TITLE
Feature/64 importance sampling MarginalCDF test

### LINKED ISSUE #64: importance sampling:  MarginalCDFExtrapolation test
# Business need
We have existing code to create importance samples but there is no test yet to ensure its quality/ accuracy.

# Definition of done
- tests for importance sampling code in MarginalCDFExtrapolation

### LINKED ISSUE #90: importance sampling test: _parameter_estimates
# Description
Part of the ticket #64. In _parameter_estimates the parameters of the individual distributions that make up the marginal distribution are estimated. As part of that the importance weights are for moved around in the function. It needs to be tested that the weights are connected still to the same samples in the outout as in the input.

# Definition of Done
- test that verifies that weights still connected to the right samples 

# Notes
- use very simple weights and samples e.g. [0,1,2,3] and [0.1,0.2,0.3]

### LINKED ISSUE #92: importance sampling test: Integration test
# Description
Full test if QoI runs with importance samples and weights.

# Definition of Done
- successful run of QoI estimation with importance samples and weights

# Notes
Similar to test_call_basic_example in test_marginal_cdf_extrapolation.py test if QoI runs with importance samples and weights -> uses deterministic GP to eliminate uncertainty introduced by using a GP

### LINKED ISSUE #95: importance sampling test: System test
# Description
Part of #64
Show that importance sampling improves QoI estimate compared to just random sampling.

### PR DESCRIPTION
**Pull request by am-kaiser**
# Business need
Importance sampling is already used in the code for MarginalCDF which is in src but there are no tests to ensure that importance sampling works as desired.

# Main issue: 
Closes #64

# Sub issues:
- closes #90
- closes #92
- closes #95 

# Notes
There will be a separate PR which tests the actual creation of the importance sampling (i.e. a test for the file importance_sampling.py which is currently part of the wave use case).

### CODE REVIEW

**Comment by am-kaiser** [comment]
I'd be especially interested in getting you impression if the comments are descriptive enough or if anything is unclear and can be better explained.

I have decided to keep the integration test (if you can call it that) even though it does not really add much but I feel like it is still nice to have a minimal test just to check if the function runs.

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    return np.ones(x.shape[0]) * 0.1
+
+
+# Ruff does not allow default arguments in test functions. Using a decorator circumvents that.
+@pytest.mark.parametrize("params", [(1, False)])
+def test_system_marginal_cdf_with_importance_sampling(
+    params: tuple[float, bool],
+) -> tuple[pd.DataFrame, float] | None:
+    """
+    Show that that the QoI estimation requires less samples with importance samples compared to when using
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:136]: nit: I don't think the test actually does this? it seems to use the same number of sample for both, but compare how much they are each able to reduce uncertainty?

It could also be worth providing a bit of a high level description of the test - something along the lines of "There is variability in the specific samples in the env_data used to instantiate a QoiEstimator, which creates the variability of the QoIEstimator  estimate. This can be seen by inspecting how the estimates given change when the QoiEstimator has been instantiated with a different dataset. This test shows how (good) importance samples can reduce the variability between estimates of QoIEstimators instantiated with different env_data. The effect of GP uncertainty is removed by using a deterministic model, meaning all uncertainty in the estimate comes from the environment samples."

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+    Args:
+        params: tuple of
+            - The error allowed in assertions is multiplied by this number.
+            - Bool to specify if the QoI results shall be returned for further testing or plotting.
+
+    Returns:
+        If return_results==True: the QoI results are returned in a pandas Dataframe
+
+    A deterministic GP is used to remove most of the uncertainty related to the GP. Some uncertainty remains as the GP
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:147]: Doesn't the deterministic GP remove all uncertainty? I don't understand this comment "Some uncertainty remains as the GP is not fit on the whole dataset." - there doesn't seem to be any fitting going on?  I would probably also put this section with the intro at the start

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    # cover the subregion where the extreme response occurs
+    env_data = np.load(Path(__file__).parent / "data" / "importance_sampling" / "environment_distribution.npy")
+
+    # Get brute force estimate
+    brute_force_path = Path(__file__).parent / "data" / "importance_sampling" / "brute_force_solution.json"
+    with brute_force_path.open() as file:
+        brute_force_qoi = json.load(file)["statistics"]["median"]
+
+    # Set up a deterministic GP which uses the true underlying functions of loc and scale used in the simulator
+    def true_underlying_func(x: torch.Tensor) -> torch.Tensor:
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:178]: `_true_scale_func` and `_true_scale_func` should probably be moved in `test_system_marginal_cdf_with_importance_sampling` if not used anywhere else. Additionally could you just make them take torch inputs and output and save yourself converting back and forth here?

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            )
+
+    jobs_output_file = None
+    qoi_results = [job(output_file=jobs_output_file) for job in qoi_jobs]
+
+    df_jobs = pd.json_normalize([item.to_dict() for item in qoi_results], max_level=1)
+    df_jobs.columns = df_jobs.columns.str.removeprefix("tags.")
+
+    # The aim of this test is to show that QoI estimation with importance sampling requires less samples
+    # (i.e. num_samples) to converge to the solution than with using the full environment dataset.
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:225]: Need to be updated similar to docstring

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+def _true_scale_func(x: NDArray[np.float64]) -> NDArray[np.float64]:
+    # For this toy example we use a constant scale for simplicity
+    return np.ones(x.shape[0]) * 0.1
+
+
+# Ruff does not allow default arguments in test functions. Using a decorator circumvents that.
+@pytest.mark.parametrize("params", [(1, False)])
+def test_system_marginal_cdf_with_importance_sampling(
+    params: tuple[float, bool],
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:133]: You can also use paramterise like this which make it a bit clearer what these variable are:
```python
@pytest.mark.parametrize("error_tolerance, return_results", [(1, False)])
def test_system_marginal_cdf_with_importance_sampling(
    error_tolerance: float,
    return_results: bool,
    ...
```

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    return np.exp(dist.log_prob(torch.tensor(x)).numpy())
+
+
+def _true_scale_func(x: NDArray[np.float64]) -> NDArray[np.float64]:
+    # For this toy example we use a constant scale for simplicity
+    return np.ones(x.shape[0]) * 0.1
+
+
+# Ruff does not allow default arguments in test functions. Using a decorator circumvents that.
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:130]: We have been using a pattern of having a parameter `visualise` which determine if the test function should show plotting. We have found this helpful as then all the plotting is contained in the test so it doesn't get lost.  It was just something we did for tests where we would live to visualize to our self what is going on when debugging - not sure if the is a cleaner way of doing it. You could consider doing something similar here?

for examples see tests/distributions/test_mixture.py: line 49 or just search "visualise" in the test directory.

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    import matplotlib.pyplot as plt
+
+    fig, ax = plt.subplots(1, 2, figsize=(10, 5), sharex=True, sharey=True)
+
+    # Plot results for full env data
+    qoi_results[qoi_results["dataset_name"] == "full"].hist(column="mean", ax=ax[0], grid=False)
+    ax[0].axvline(brute_force_qoi, c="orange", label=f"Brute force ({brute_force_qoi:.2f})")
+    ax[0].set_title(
+        "Dataset: full, "
+        "mean={qoi_results.loc[qoi_results['dataset_name'] == 'full', 'mean'].mean():.2f}, "
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:267]: I think these are all mean to be f strings? currently they are going to print "{qoi_results.loc[qoi_results['dataset_name'] == 'full', 'mean'].mean():.2f}"

**Review by swinter1 — CHANGES_REQUESTED** [review]
Don't be put off by the number of comments, they are mainly cosmetic stuff - you've made a good start here, nice work.

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    # Get posterior samples and connected weights for given QoI estimator
+    posterior_samples, importance_weights_qoi = qoi_estimator._parameter_estimates(gp_passthrough_1p)
+
+    # As a deterministic GP is used the posterior samples should not only have the same values as the
+    # input samples but their order should be the same as well
+    assert torch.equal(torch.flatten(importance_samples), posterior_samples[..., 0][0])
+    assert torch.equal(torch.flatten(importance_weights), torch.flatten(importance_weights_qoi))
+
+
+def test_qoi_runs_with_importance_sampling(gp_passthrough_1p: GenericDeterministicModel):
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:79]: I was also on the fence about keeping this function - I'm leaning towards removing it. Reason being is `_parameter_estimates` always returns weights of a given shape. We have other tests ensuring the `_parameter_estimates` returns the correct output for both importance and non-importance version - so this function doesn't do anything different for the different cases.

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py (hunk 1 above)
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py]: true! I had removed the convergence part from the test and only kept the reduction of uncertainty part but forgot to change the docstring.

Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py]: done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py (hunk 4 above)
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py]: done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py
[the remark this answers is not in the document]
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:1]: I was considering that but then it would make most sense if test_parameter_estimates_consistency_of_weights is moved to test_marginal_cdf_extrapolation.py and test_system_marginal_cdf_with_importance_sampling to test_marginal_cdf_extrapolation_system.py. However, test_marginal_cdf_extrapolation_system.py is already very long and hard to understand. So I thought it would be better to not further add to it and have the importance sampling part in a separate test file to make it easier to understand.

[the remark this answers is not in the document]
Comment (swinter1) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:1]: hmmm - I get your point. What I want to steer clear of is getting a new test file for each different component of a tool we test. 2 potential options:
- Perhaps the rule we can follow is unit/integration tests all go in the standard file, and system tests can go it their own file? e.g have `test_marginal_cdf_extrapolation_system_importance_sampling`.
- Say that systems tests should also just go in the standard file, and then make a todo/ticket to shift `test_marginal_cdf_extrapolation_system.py.` That file should probably be cleaned up at some point anyway.

What are your thoughts?

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        torch.cat([batch[0] for batch in env_and_importance_data]).flatten(), posterior_samples[..., 0][0]
+    )
+    assert torch.equal(
+        torch.cat([batch[1] for batch in env_and_importance_data]).flatten(), torch.flatten(importance_weights_qoi)
+    )
+
+
+# Ruff does not allow default arguments in test functions. Using a decorator circumvents that.
+@pytest.mark.parametrize("error_tolerance, visualize", [(1, False)])
+def test_system_marginal_cdf_with_importance_sampling(
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:90]: Test also need to be marked `non-determinstic` and `system`. See `test_marginal_cdf_extrapolation_system.py` for how this is done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    df_jobs.columns = df_jobs.columns.str.removeprefix("tags.")
+
+    # The aim of this test is to show that QoI estimation with importance sampling results in less uncertainty in the
+    # QoI compared to when using the full environment dataset.
+    # All absolute values in the following asserts are chosen based on visual inspection of the QoI results.
+    # There are two criteria we use to judge the convergence:
+    # 1. the mean of the QoI means is close to the brute force solution and in particular the one with
+    # importance sampling is closer than the one without
+    assert (
+        abs(df_jobs.loc[df_jobs["dataset_name"] == "importance_sample", "mean"].mean() - brute_force_qoi)
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:195]: Please use a variable rather than repeat this. Also make a lighter mental load for reader to understand what this statement does, will also make the plotting nicer to read.

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    qoi_results = [job(output_file=jobs_output_file) for job in qoi_jobs]
+
+    df_jobs = pd.json_normalize([item.to_dict() for item in qoi_results], max_level=1)
+    df_jobs.columns = df_jobs.columns.str.removeprefix("tags.")
+
+    # The aim of this test is to show that QoI estimation with importance sampling results in less uncertainty in the
+    # QoI compared to when using the full environment dataset.
+    # All absolute values in the following asserts are chosen based on visual inspection of the QoI results.
+    # There are two criteria we use to judge the convergence:
+    # 1. the mean of the QoI means is close to the brute force solution and in particular the one with
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:192]: I think when we spoke about this we came to the conclusions that the importance sample version is not necessarily expected to have a better mean estimate  - the only thing we can expect is that it has lower variance than the other method. Has this understanding changed? if not -  i think the comments and tests should reflect that - especially as this test doubles as a mini tutorial to help people get a better understanding

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    )
+
+    assert abs(df_jobs.loc[df_jobs["dataset_name"] == "importance_sample", "mean"].mean() - brute_force_qoi) <= abs(
+        df_jobs.loc[df_jobs["dataset_name"] == "full", "mean"].mean() - brute_force_qoi
+    )
+
+    # 2. the std of the QoI means is small with importance sampling
+    assert df_jobs.loc[df_jobs["dataset_name"] == "importance_sample", "mean"].std() <= 0.25 * error_tolerance
+
+    if visualize:
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:206]: Good plots. If we think of this test as also a mini demo - would it be helpful/a lot of work to plot the problem (e.g the importance sampled and standard env distribution, and the response surface)? Do you think it is worth the effort?

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+    # The environment is chosen such that it is concentrated in one region of the search space and only few samples
+    # cover the subregion where the extreme response occurs
+    env_data = np.load(Path(__file__).parent / "data" / "importance_sampling" / "environment_distribution.npy")
+
+    # Get brute force estimate
+    brute_force_path = Path(__file__).parent / "data" / "importance_sampling" / "brute_force_solution.json"
+    with brute_force_path.open() as file:
+        brute_force_qoi = json.load(file)["statistics"]["median"]
+
+    # Set up a deterministic GP which uses the true underlying functions of loc and scale used in the simulator
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:133]: done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+    def _true_underlying_func(x: torch.Tensor) -> torch.Tensor:
+        # For this toy example we use a MultivariateNormal distribution
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:135]: done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py (hunk 11 above)
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:192]: done

File: tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py
[the remark this answers is not in the document]
Comment (am-kaiser) [reply tests/qoi/test_marginal_cdf_extrapolation_with_importance_sampling.py:1]: done

File: tests/qoi/test_marginal_cdf_extrapolation.py [hunk 15]
```diff
@@ -159,6 +175,193 @@ def test_parameter_estimates_env_batch_invariant(
...
+            dtype=torch.float64,
+        )
+
+        # Get posterior samples and connected weights for given QoI estimator
+        posterior_samples, importance_weights_qoi = qoi_estimator._parameter_estimates(gp_passthrough_1p)
+
+        # As a deterministic GP is used the posterior samples should not only have the same values as the
+        # input samples but their order should be the same as well
+        assert torch.equal(torch.tensor([1.0, 2.0, 3.0, 4.0]), posterior_samples[..., 0][0])
+        assert torch.equal(torch.tensor([0.1, 0.2, 0.3, 0.4]), torch.flatten(importance_weights_qoi))
```
Comment (swinter1) [inline tests/qoi/test_marginal_cdf_extrapolation.py:221]: I think we wanted to also check that the whole output had the correct dimensions - thats why we added those dimension checks originally. If we explicitly make the full output we are expecting that should be easy to do and informative for the reader.

**Review by swinter1 — APPROVED** [review]
